### PR TITLE
esp32_can: make queue lengths configurable

### PR DESCRIPTION
**Pull request by mrk-its**
# What does this implement/fix?

It adds optional parameters `rx_queue_len` and `tx_queue_len` to esp32_can config. esp32 twai driver assumes length 5 by default which is pretty small value and frequently leads to packet loss. This is change is backward compatible: when these parameters are not set, then esp32 twai default will be used.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4199

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
canbus:
  - platform: esp32_can
    # ...
    tx_queue_len: 128
    rx_queue_len: 128
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/esp32_can/canbus.py [hunk 1]
```diff
@@ -87,3 +94,7 @@ async def to_code(config):
 
     cg.add(var.set_rx(config[CONF_RX_PIN]))
     cg.add(var.set_tx(config[CONF_TX_PIN]))
+    if CONF_RX_QUEUE_LEN in config:
+        cg.add(var.set_rx_queue_len(config[CONF_RX_QUEUE_LEN]))
```
Comment (jesserockz) [inline esphome/components/esp32_can/canbus.py]:
```suggestion
    if (rx_queue_length := config.get(CONF_RX_QUEUE_LEN, None)) is not None:
        cg.add(var.set_rx_queue_len(rx_queue_length))
```

File: esphome/components/esp32_can/esp32_can.h [hunk 2]
```diff
@@ -12,6 +12,8 @@ class ESP32Can : public canbus::Canbus {
  public:
   void set_rx(int rx) { rx_ = rx; }
   void set_tx(int tx) { tx_ = tx; }
+  void set_tx_queue_len(int tx_queue_len) { tx_queue_len_ = optional<int>(tx_queue_len); }
+  void set_rx_queue_len(int rx_queue_len) { rx_queue_len_ = optional<int>(rx_queue_len); }
```
Comment (jesserockz) [inline esphome/components/esp32_can/esp32_can.h]: Please use the correct types
```suggestion
  void set_tx_queue_len(uint32_t length) { this->tx_queue_len_ = length; }
  void set_rx_queue_len(uint32_t length) { this->rx_queue_len_ = length; }
```

File: esphome/components/esp32_can/esp32_can.h [hunk 3]
```diff
@@ -21,6 +23,8 @@ class ESP32Can : public canbus::Canbus {
 
   int rx_{-1};
   int tx_{-1};
+  optional<int> tx_queue_len_{};
+  optional<int> rx_queue_len_{};
```
Comment (jesserockz) [inline esphome/components/esp32_can/esp32_can.h]: Please use the correct types
```suggestion
  optional<uint32_t> tx_queue_len_{};
  optional<uint32_t> rx_queue_len_{};
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/esp32_can/esp32_can.h (hunk 2 above)
Comment (mrk-its) [reply esphome/components/esp32_can/esp32_can.h]: good catch, thanks!

File: esphome/components/esp32_can/esp32_can.h [hunk 4]
```diff
@@ -12,6 +12,8 @@ class ESP32Can : public canbus::Canbus {
  public:
   void set_rx(int rx) { rx_ = rx; }
   void set_tx(int tx) { tx_ = tx; }
+  void set_tx_queue_len(uint32_t tx_queue_len) { tx_queue_len_ = tx_queue_len; }
```
Comment (jesserockz) [inline esphome/components/esp32_can/esp32_can.h]: > All uses of class members and member functions should be prefixed with this-> to distinguish them from global functions/variables.
https://esphome.io/guides/contributing#codebase-standards

```suggestion
  void set_tx_queue_len(uint32_t tx_queue_len) { this->tx_queue_len_ = tx_queue_len; }
```

File: esphome/components/esp32_can/esp32_can.h [hunk 5]
```diff
@@ -12,6 +12,8 @@ class ESP32Can : public canbus::Canbus {
...
+  void set_rx_queue_len(uint32_t rx_queue_len) { rx_queue_len_ = rx_queue_len; }
```
Comment (jesserockz) [inline esphome/components/esp32_can/esp32_can.h]:
```suggestion
  void set_rx_queue_len(uint32_t rx_queue_len) { this->rx_queue_len_ = rx_queue_len; }
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Comment by mrk-its** [comment]
@jesserockz I created documentation PR already: esphome/esphome-docs#4199, forgot to update the PR desccription :facepalm:

**Pull request merged**